src/consumer/runner.js

```diff
@@ -295,13 +295,14 @@ module.exports = class Runner extends EventEmitter {
       })
     }
 
-    const { lock, unlock, unlockWithError } = barrier()
+    const { lock, unlock } = barrier()
     const concurrently = limitConcurrency({ limit: this.partitionsConsumedConcurrently })
 
     let requestsCompleted = false
     let numberOfExecutions = 0
     let expectedNumberOfExecutions = 0
     const enqueuedTasks = []
+    let error = null // the first fetch request, processing or heartbeat error encountered
```

Comment:
Instead of keeping this error reference and making sure that we catch everywhere that can throw, we can query the state of the `lock`. If it is unlocked, it means that we should skip any further processing.

We'd just need to expose the lock state from the barrier:

```js
module.exports = () => {
  let unlock
  let unlockWithError
  let locked = true
  const lock = new Promise(resolve => {
    unlock = () => {
      locked = false
      resolve()
    }
    unlockWithError = error => {
      locked = false
      resolve(error)
    }
  })

  return { lock, locked, unlock, unlockWithError }
}
```

Then when we want to check if we should skip the batch, instead of checking for `error` we check for `locked`.

There are two benefits from this:
1. We don't need this additional reference to the error, since this is already kept in the lock. Less moving parts to keep track of.
2. If we miss updating the error reference somewhere that could throw, this will still be handled by the `.catch` on each `enqueuedTask`

Reply:
But the point is that we don't want to unlock the barrier as soon as we get an error.  If we do that then the barrier promise and thus the fetch() promise settle before any already started batch processing has completed.
We could move the error state to the barrier.  I'll try that - let's see if its clearer.

Reply:
@Nevon I have re-worked it to something I think shows more clearly what is going on.  Let me know if you agree.  I'm aware the tests still need improvements.

Reply:
I understand your point now. So you want to keep the lock so that we wait for all processing to complete, but you need to keep a reference to the error so that it can be thrown at the end. I think it makes sense.

The one thing I'm wondering about is whether or not we can have multiple errors - since currently we're only keeping a single reference. Perhaps `setError` should be wrapping all the errors [in an AggregateError](https://github.com/tulios/kafkajs/blob/master/src/errors.js#L235-L241) so that we don't lose information.

Reply:
@Nevon sorry, seems I didn't push.  Anyway, I think we both understanding it the same now.
The changes I just pushed are in line with what you suggest - we collect all the errors.  I'll add to it to throw the aggregateError (instead of just the first) asap.  And update test cases.

Reply:
Re-worked to always throw an AggregateError from fetch().